consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,171 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+// If this node is a proxy, it will use the most recent directAnnounce this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
+	shouldSend, err := sb.shouldSendValidatorProof(peer)
+	if err != nil {
+		return nil, err
+	}
+	if shouldSend {
+		if sb.config.Proxy {
+			// if this proxy has been disconnected from its validator for any reason,
+			// don't send a proof message, but do not fail the handshake.
+			if sb.proxyDirectAnnounceMsg != nil && sb.proxiedPeer != nil {
+				// Make a copy of the proxyDirectAnnounceMsg for thread safety
+				sb.proxyDirectAnnounceMsgMu.RLock()
+				defer sb.proxyDirectAnnounceMsgMu.RUnlock()
+				return sb.proxyDirectAnnounceMsg.Copy(), nil
+			}
+		} else {
+			return sb.generateDirectAnnounce(getCurrentAnnounceVersion())
+		}
+	}
+	// Even if we decide not to identify ourselves,
+	// send an empty message to complete the handshake
+	return &istanbul.Message{}, nil
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+	if !regAndActiveVals[validatorAddress] {
+		return false, nil
+	}
+	// If the peer is not a known validator, we do not want to give up extra
+	// information about ourselves
+	if _, err := sb.valEnodeTable.GetAddressFromNodeID(peer.Node().ID()); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Msg is empty, the peer has decided not to reveal its info
+	if len(msg.Msg) == 0 {
+		return false, nil
+	}
+
+	var directAnnounce directAnnounce
+	err = rlp.DecodeBytes(msg.Msg, &directAnnounce)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the peer is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		sb.logger.Trace("Error in retrieving registered/elected valset", "err", err)
+		return false, err
+	}
+	if !regAndActiveVals[msg.Address] {
+		return false, nil
```

Comment:
nit: Should we return an error here? Or log something? IIUC, with everyone running the proper code, we should never, or almost never, reach this line